### ISSUE TITLE
ttnn.gelu (accurate variant, float32): the rewritten FP32 branch still loses the negative tail to `1 + erf` cancellation — non-monotonic output, up to 19.6M representable floats from torch and 700 % from float64 for x < −4.75 (simulator measurement)

### ISSUE THREAD

**Issue by beyondpremiere** [tenstorrent/tt-metal#56490]
Verified against tt-metal `main` @ `70f840f46adb33ec21ec8893f48ee06692648f75` (2026-09-14).

**Component / Area:** kernels (SFPU) / TTNN eltwise unary
**Issue Type:** Bad Outputs

## Summary

`calculate_gelu<APPROXIMATION_MODE=false, is_fp32_dest_acc_en=true>` evaluates
`x * (0.5f + 0.5f * erf_val)` (`ckernel_sfpu_gelu.h:318`, identical on wormhole_b0 and
blackhole). For x < −3.8 the sum `1 + erf(x/√2)` cancels catastrophically, so the output
is dominated by rounding noise of the erf fit: it is non-monotonic where GELU is strictly
monotone (as is torch float32, to a lesser degree), and its error vs float64 GELU grows from 3 % (max on [−4.5, −4.0]) to 33 %
([−5.0, −4.5]), 131 % ([−5.25, −5.0]) and 700 % ([−5.5425, −5.25], endpoint included). This is the branch
`ttnn.gelu` takes by default for float32 tensors. Measured on the ttsim functional
simulator (Blackhole) through the tt-llk kernel harness, not on silicon and not through the
ttnn API — see Scope.

## Observed

Dense sweep, 2048 float32 inputs evenly spaced over [−5.5425, −3.0], exact-mode GELU,
fp32 in / fp32 out / fp32 dest (`unpack_to_dest=true`). Reference is
`0.5·x·erfc(−x/√2)` in float64 on the exact float32 input (the stable form; it differs
from the `1+erf` float64 form by < 2e-9 relative here, so the float64 reference is not the
source of any number below).

| x (fp32) | kernel | torch float32 | float64 | kernel↔torch (representable floats) | kernel vs f64 | torch vs f64 |
|---|---|---|---|---|---|---|
| −3.0 | −4.049689e-03 | −4.049689e-03 | −4.049694e-03 | 0 | 0.00 % | 0.00 % |
| −3.999860 | −1.265958e-04 | −1.267150e-04 | −1.267557e-04 | 8,191 | 0.13 % | 0.03 % |
| −4.249514 | −4.571899e-05 | −4.546569e-05 | −4.551967e-05 | 69,624 | 0.44 % | 0.12 % |
| −4.500411 | −1.515586e-05 | −1.528999e-05 | −1.526142e-05 | 130,319 | 0.69 % | 0.19 % |
| −4.750065 | −4.813141e-06 | −4.813141e-06 | −4.829662e-06 | 0 | 0.34 % | 0.34 % |
| −4.999719 | −1.341029e-06 | −1.490032e-06 | −1.435266e-06 | 1,310,647 | 6.57 % | 3.82 % |
| −5.250616 | −1.564805e-07 | −4.694416e-07 | −3.979750e-07 | 13,894,277 | 60.68 % | 17.96 % |
| −5.399663 | −4.827675e-07 | −1.609225e-07 | −1.802575e-07 | 13,946,238 | 167.82 % | 10.73 % |
| −5.500270 | −1.639208e-07 | −1.639208e-07 | −1.042880e-07 | 0 | 57.18 % | 57.18 % |
| −5.5425 | −6.607175e-07 | −1.651794e-07 | −8.263407e-08 | 16,777,216 | 699.57 % | 99.89 % |

These are sample rows, not thresholds. Maxima per interval over all sampled points:

| interval | n | kernel max rel err vs f64 | torch fp32 max rel err vs f64 | kernel points > 1 % | > 10 % | max kernel↔torch (representable floats) |
|---|---|---|---|---|---|---|
| [−3.5, −3.0] | 403 | 0.04 % | 0.01 % | 0 % | 0 % | 7,108 |
| [−4.0, −3.5] | 403 | 0.29 % | 0.04 % | 0 % | 0 % | 32,248 |
| [−4.5, −4.0] | 402 | 3.21 % | 0.41 % | 9.0 % | 0 % | 363,587 |
| [−5.0, −4.5] | 403 | 33.03 % | 4.51 % | 75.7 % | 12.4 % | 5,154,320 |
| [−5.25, −5.0] | 201 | 131.09 % | 17.16 % | 95.5 % | 74.6 % | 16,777,216 |
| [−5.5425, −5.25] | 236 | 699.57 % | 99.89 % | 99.2 % | 88.1 % | 19,590,612 |

Monotonicity: GELU is strictly decreasing as x increases on this interval (x·Φ(x) → 0⁻
as x → −∞; the float64 reference runs from −8.26e-08 at −5.5425 down to −4.05e-03 at −3.0
with no adjacent-sample increase). Both the kernel output and torch float32 are
non-monotonic here — each has hundreds of adjacent samples where the output increases
although GELU decreases (counts are in the CSV; they are not used to argue severity, which
rests on the error magnitudes above).

Worst case: x = −5.36612749 (`0xc0abb751`): kernel −7.99615293e-07 (`0xb556a525`),
torch float32 −1.59923061e-07 (`0xb42bb751`), float64 −2.158393033e-07. The two fp32
outputs are 19,590,612 representable floats apart; measured as |Δ| / fp32 spacing at the
torch value the same point reads 45,014,339 (that metric inflates when the outputs straddle
exponent ranges, which is why the count of representable floats is used above).

The core region [−3, 5] is fine on the same build (max 712 spacing-ULP vs float64 over a
2048-point ramp; ≤ 1 for most of it).

Metric definitions: "representable floats" = |ordinal(a) − ordinal(b)| over float32 bit
patterns; relative error = |kernel − f64| / |f64|.

## Expected

x · Φ(x) (`torch.nn.functional.gelu(x, approximate="none")` semantics). torch's own float32
evaluation uses the same `1 + erf` form and therefore also degrades in the deep tail (see
the torch columns), so "matches torch float32" is not an adequate acceptance criterion
here; a float64 reference on the exact fp32 inputs is.

## Mechanism

`tt_metal/hw/ckernels/{wormhole_b0,blackhole}/metal/llk_api/llk_sfpu/ckernel_sfpu_gelu.h`
— the two copies are identical for this function, same line numbers:

- `:299` — `else if constexpr (is_fp32_dest_acc_en)`: the FP32 accurate branch.
- `:304` — `GELU_SAT = -5.54259443f`; output is 0 at or below it.
- `:318` — `result = x * (0.5f + 0.5f * erf_val);` with `erf_val` from the n16/d16
  rational erf (`GELU_ERF_NUM/DEN`, `:258`), clamped to [−1, 1].
- `:322` — `v_if(erf_val == -1.0f) { result = x * 2^-25; }` (stuck-erff guard).
- `:335` — the **bf16** branch instead calls `calculate_gelu_piecewise` (`:150`), whose
  tail region (−5.54259443, −3.125] computes `H ≈ ½·erfc(|x|/√2)` directly as
  Moroz `exp_21f(−x²/2)` × a degree-3 correction polynomial (`GELU_HCORR_*`), then snaps
  H to a 2⁻²⁵ grid (`ROUND_TO_GRID`, `:170-180`) to reproduce torch's float32 staircase
  at bf16 resolution.

At x = −5, erf(x/√2) = −0.999999427, so the quantity the kernel actually forms,
`0.5f + 0.5f * erf_val` ≈ 2.87e-07, is about five float32 spacings (5.96e-08) at magnitude
0.5 — the sum retains two or three significant bits regardless of the erf fit's quality.
The rational erf is accurate to ~3.5 fp32 spacing-ULP on this build (measured separately;
absolute error ≈ 2e-07, i.e. ≈ 1e-07 after the 0.5 scaling), which is ~35 % of that
quantity at x = −5 and exceeds it near the saturation point. That reproduces the error
profile above. Both tested implementations lose accuracy in this interval, since both
evaluate Φ on the cancelling side of the identity; the kernel has the larger measured
maximum relative error in every reported interval. No claim is made about the rounding
quality of PyTorch's intermediate erf.

The branch is the one `ttnn.gelu` uses for float32 input:
`ttnn/cpp/ttnn/operations/eltwise/unary/unary.cpp:56-60` sets `fp32_dest_acc_en = true`
when the input dtype is FLOAT32 (`preserve_fp32_precision`), and `unary.hpp` defaults
`gelu(...)` to `GeluVariant::ACCURATE` (`APPROXIMATION_MODE=false`).

## Relation to earlier reports

- #35290 (Jan 2026, closed Apr 2026) reported catastrophic ULP in `ttnn.gelu` accurate mode
  and named erf saturation → `1 + (−1)` cancellation as the root cause (for the fused
  `gelu_bw`, #35971). #46185 (closed Jun 17) then replaced the FP32 path with the current
  erf-rational branch. #32970 (open, TT tracking) lists `aten.gelu.default` BH fp32 at
  > 70M ULP from before that rewrite. #52407 (open) lists `fp32_dest_acc` as unswept.
- This report is about the **residual in the rewritten branch at `70f840f`**: the
  formulation at `:318` still evaluates Φ on the cancelling side, so the tail defect
  survived the rewrite. I did not find a report of that residual; I am not asserting there
  is none.
- `gelu_bw` has separate open reports on its fp32 tail (#54867, #54883); those are a
  different kernel (`calculate_gelu_derivative_*`) and are not covered here.

## Proposed direction (not a validated implementation)

Evaluate Φ(x) for x < −3.125 on the non-cancelling side, i.e. as ½·erfc(−x/√2) or as
exp(−x²/2)·corr(x), and keep the existing erf-based core for x ≥ −3.125 and the
saturation at −5.54259443.

Two constraints on how, from the pinned source:

1. There is **no fp32-accurate erfc fit to reuse** at this revision.
   `ckernel_sfpu_erfc.h` is still the two-segment n4/d5 rational whose own header says
   "FP32 MaxULP≈9M" (`:23-30`); #54053 requested an accurate path but none has landed.
   An erfc-based fix therefore needs its own tail fit (erfc on [2.2, 3.92] is smooth and
   well conditioned; a low-degree rational or exp×polynomial suffices) — that fit is new
   work to be validated, not an existing asset.
2. The bf16 branch's tail (`calculate_gelu_piecewise`, `:150-181`) already has the right
   structure — `exp_21f(−x²/2)` × correction polynomial, no subtraction — but its
   correction polynomial and its 2⁻²⁵ grid snap are designed to reproduce torch's bf16
   result, not to reach fp32 accuracy against the true function. Reusing the structure in
   the fp32 branch requires an fp32-grade correction fit and dropping/adjusting the grid
   snap; the resulting accuracy (target: a few fp32 ULP, monotone) is a **validation
   goal**, not a property the existing branch is known to have. No performance claim is
   made either way.

## Tests

- Add an fp32 tail sweep (≥ 2048 points over [−5.5425, −3.0], `is_fp32_dest_acc_en=true`)
  to `tt_metal/tt-llk/tests/python_tests/test_eltwise_unary_sfpu.py` asserting
  (a) no adjacent-sample increase of the output and (b) a bound in representable floats against
  `0.5·x·erfc(−x/√2)` evaluated in float64 on the exact fp32 inputs — not against torch
  float32.
- Extend the ttnn `gelu` sweep with a float32 tail range so the branch float32 inputs
  actually take is exercised end to end.

## Reproduction (as performed)

- tt-metal `70f840f46adb33ec21ec8893f48ee06692648f75`, tt-llk harness
  (`tt_metal/tt-llk/tests/python_tests`), SFPI 7.76.0 (`riscv-tt-elf-g++ 15.1.0`, build
  929), tt-exalens 0.3.31, torch 2.9.1+cpu, numpy 2.2.6.
- Simulator: ttsim v1.10.7 `libttsim_bh_aarch64.so` (sha256
  `12c24c0bd11b3d582c1e7c6dc775c17e30e3b52c774450617b20fd53f73597e3`), `soc_descriptor.yaml`
  = `blackhole_140_arch.yaml`; `TT_METAL_SIMULATOR` set to it,
  `TT_METAL_DISABLE_SFPLOADMACRO=1`; `pytest --run-simulator --forked`.
- Kernel/config: `TestConfig("sources/eltwise_unary_sfpu_test.cpp", Float32→Float32,
  MATH_OP=Gelu, APPROX_MODE=No, FAST_MODE=No, CLAMP_NEGATIVE=True, dest_acc=Yes,
  unpack_to_dest=True)`, `src_A = torch.linspace(-5.5425, -3.0, 2048)`, input dims
  [32, 64]. `test_gelu_fp32_tail.py` (Files below) is the exact file; it writes the
  CSV (input/output bit patterns, torch float32, float64 reference; version header on
  line 1).
- Note for reproducers: the tt-llk harness does not define `INP_FLOAT32`; that affects erf,
  i1, digamma and softplus (reported in #56492) but **not** this GELU branch, which keys on
  `is_fp32_dest_acc_en` and produced bit-identical output with and without the define.

## Scope

- Simulator measurement (Blackhole ttsim), through the tt-llk kernel harness with the same
  template configuration ttnn selects for float32 input. Not a `ttnn.gelu` API call, not
  silicon. The Wormhole copy of the function is byte-identical at this revision; the
  Wormhole simulator currently rejects the `unpack_to_dest` configuration (reported as
  tenstorrent/ttsim#18), so WH is covered by code identity, not execution.
- The absolute magnitudes involved are small (|GELU| < 1.5e-6 for x < −5). The report's
  claim is about correctness of the FP32 path relative to fp32 precision and monotonicity,
  not about model-level impact, which is not assessed here.
- Not claiming the erf fit is defective (~3.5 ULP); the defect is the formulation at
  `:318`.

## Files

- Reproducer and the 2048-row CSV (input/output float32 bit patterns, torch float32, float64 `erfc`-form reference; version header on line 1): https://gist.github.com/beyondpremiere/1bcdde9a510860fd06b35bb28ec4cfa8
- The reproducer inline, for review without leaving the page:

<details>
<summary>test_gelu_fp32_tail.py</summary>

```python
# Dense fp32 sweep of exact GELU over the negative tail (-5.5425, -3.0], where
# calculate_gelu<false, is_fp32_dest_acc_en=true> computes x*(0.5+0.5*erf(x/sqrt2)) and
# 1+erf cancels catastrophically. Modeled on test_sqrt_custom_infinity_regression.
import csv, math, os
import pytest, torch
from helpers.format_config import DataFormat, InputOutputFormat
from helpers.golden_generators import TILE_DIMENSIONS
from helpers.llk_params import (ApproximationMode, BlocksCalculationAlgorithm, DestAccumulation, FastMode, MathOperation)
from helpers.param_config import get_num_blocks_and_num_tiles_in_block
from helpers.stimuli_config import StimuliConfig
from helpers.test_config import TestConfig
from helpers.test_variant_parameters import (APPROX_MODE, CLAMP_NEGATIVE, FAST_MODE, MATH_OP, NUM_BLOCKS, NUM_TILES_IN_BLOCK, TILE_COUNT, DestSync, generate_input_dim)

@pytest.mark.parametrize("lo,hi", [(-5.5425, -3.0)])
def test_gelu_fp32_tail(lo, hi):
    formats = InputOutputFormat(DataFormat.Float32, DataFormat.Float32)
    dest_acc = DestAccumulation.Yes
    input_dimensions = [32, 64]  # 2048 elements = 2 tiles
    n = input_dimensions[0] * input_dimensions[1]
    src_A = torch.linspace(lo, hi, n, dtype=torch.float32)
    src_B = torch.zeros(n, dtype=torch.float32)
    tile_cnt = (input_dimensions[0] // 32) * (input_dimensions[1] // 32)
    num_blocks, num_tiles_in_block = get_num_blocks_and_num_tiles_in_block(
        DestSync.Half, dest_acc, formats, input_dimensions, TILE_DIMENSIONS, BlocksCalculationAlgorithm.Standard)
    cfg = TestConfig(
        "sources/eltwise_unary_sfpu_test.cpp", formats,
        templates=[generate_input_dim(input_dimensions, input_dimensions), APPROX_MODE(ApproximationMode.No),
                   FAST_MODE(FastMode.No), CLAMP_NEGATIVE(True), MATH_OP(mathop=MathOperation.Gelu)],
        runtimes=[TILE_COUNT(tile_cnt), NUM_BLOCKS(num_blocks), NUM_TILES_IN_BLOCK(num_tiles_in_block)],
        variant_stimuli=StimuliConfig(src_A, formats.input_format, src_B, formats.input_format, formats.output_format,
                                      tile_count_A=tile_cnt, tile_count_B=tile_cnt, tile_count_res=tile_cnt),
        dest_acc=dest_acc, unpack_to_dest=True)
    res = torch.tensor(cfg.run().result, dtype=torch.float32)
    out = os.environ.get("GELU_TAIL_CSV", "/work/gelu_fp32_tail.csv")
    import struct, subprocess
    bits = lambda v: "0x%08x" % struct.unpack("<I", struct.pack("<f", float(v)))[0]
    sha = subprocess.run(["git", "-C", "/opt/tt-metal", "rev-parse", "HEAD"], capture_output=True, text=True).stdout.strip()
    with open(out, "w", newline="") as f:
        f.write(f"# tt-metal {sha}; torch {torch.__version__}; sfpi {open('/opt/tt-metal/tt_metal/tt-llk/tests/sfpi/sfpi.version').read().strip()}; "
                f"TT_METAL_SIMULATOR={os.environ.get('TT_METAL_SIMULATOR')}; kernel sources/eltwise_unary_sfpu_test.cpp MATH_OP=Gelu APPROX=No "
                f"Float32->Float32 dest_acc=Yes unpack_to_dest=True CLAMP_NEGATIVE=True; reference = 0.5*x*erfc(-x/sqrt2) in float64 on the exact fp32 input\n")
        w = csv.writer(f); w.writerow(["x", "x_bits", "hw", "hw_bits", "torch_f32", "torch_bits", "f64_erfc"])
        for xv, hv in zip(src_A.tolist(), res.tolist()):
            t32 = torch.nn.functional.gelu(torch.tensor(xv, dtype=torch.float32)).item()
            f64 = 0.5 * xv * math.erfc(-xv / math.sqrt(2.0))
            w.writerow([f"{xv:.9g}", bits(xv), f"{hv:.9g}", bits(hv), f"{t32:.9g}", bits(t32), f"{f64:.10g}"])
    assert len(res) == n
```

</details>

**Comment by beyondpremiere** [comment]
Related but independent: #56517 reports the same class of failure in the *tanh* GELU variant (`calculate_gelu_tanh`, same file), where the cancelling quantity is `1 + tanh(u)` rather than `0.5 + 0.5*erf`. Different kernel function, different expression, separate fix — noting it here so the two are not read as duplicates.
